UM/Settings/Setting.py

```diff
@@ -352,28 +329,103 @@ def getAllChildren(self):
     def getChildren(self):
         return self._children
 
+    def parseValue(self, value):
+        if not type(value) is str:
+            return value
+
+        if self._type == "bool":
+            return bool(ast.literal_eval(value))
+        elif self._type == "int":
+            return int(ast.literal_eval(value))
+        elif self._type == "float":
+            return float(ast.literal_eval(value.replace(",", ".")))
+        elif self._type == "string" or self._type == "enum":
+            return value
+        else:
+            return ast.literal_eval(value)
+
     def __repr__(self):
         return "<Setting: %s>" % (self._key)
 
-## private:
+    def __deepcopy__(self, memo):
+        copy = Setting(
+            self._key,
+            self._i18n_catalog,
+            label = self._label,
+            type = self._type
+        )
+
+        copy._description = self._description
+        copy._default_value = self._default_value
+        copy._visible = self._visible
+        copy._category = self._category
+        copy._unit = self._unit
+        copy._validator = deepcopy(self._validator, memo)
+        copy._parent = self._parent
+        copy._hide_if_all_children_visible = self._hide_if_all_children_visible
+        copy._children = deepcopy(self._children, memo)
+        copy._options = deepcopy(self._options, memo)
+        copy._inherit = self._inherit
+        copy._inherit_function = deepcopy(self._inherit_function, memo)
+
+        copy._warning_description = self._warning_description
+        copy._error_description = self._error_description
+        copy._enabled = deepcopy(self._enabled, memo)
+
+        copy._fixChildren()
+
+        return copy
+
+    ## private:
+
+    # Fixes parent assignment after a deepcopy operation
+    def _fixChildren(self):
+        for child in self._children:
+            child.setParent(self)
+            child._fixChildren()
+
     def _onChildVisibileChanged(self, setting):
         self.visibleChanged.emit(setting)
         self.visibleChanged.emit(self)
 
-    # Create a function that will run \param code, making the names in \param names available as local variables
-    def _createInheritFunction(self, code, names):
-        def inherit(parent, settings, c = code, n = names):
+    def _createFunction(self, code):
+        try:
+            tree = ast.parse(code, "eval")
+            names = _SettingExpressionVisitor().visit(tree)
+            compiled = compile(code, self._key, "eval")
+        except (SyntaxError, TypeError) as e:
```

Comment:
Are we sure these are all the exceptions that can occur? This is the kind of stuff that people will tweak, so good error checking is important.